### ISSUE TITLE
Goals of this repository?

### ISSUE THREAD

**Issue by fingolfin** [gap-system/containers#1]
Let's collect some goals for this repository, to see if we can agree on a common vision; then next we can see how to implement them and who will do it. CC @ZachNewbery @alex-konovalov @james-d-mitchell @wilfwilson please feel free to edit this and/or to comment on it all

- merge all existing GAP docker repositories into this one
- thus, provide say "nightly" docker images for gap-system/gap branches...
   - `master`
   - `stable-4.x` (only update if there really were changes? what if the base image changes???) 
- but also provide variants of the docker images based on GAP releases (and using the release tarballs, *not* some git snapshots
   - we'll want these under tags matching the GAP tags, so e.g. `v4.11.1` and `v4.11.0`
   - but ideally also something to say "give me the latest 4.11.x release" -- say via tag `v4.11`
   - perhaps also a `latest` tag which maps to whatever is the latest GAP release
- besides the GAP version/branch, other interesting variants include:
  - a "full" image which has all package compiled (and hence provides "all" their dependencies, or at least as close as we can get
  - a "minimal" image which has GAP, and perhaps a few select packages compiled (say anything w/o non-default external dependencies can be compiled: io, cvec, etc.; but not something like `NormalizInterface`)
  - .... more?

Some further ideas (should perhaps be moved to separate issues):

- look into images that are not based on Ubuntu but rather something more lightweight: e.g. [Alpine Linux](https://www.alpinelinux.org)
- ...

**Comment by wilfwilson** [comment]
Thanks for posting this issue. I will come back to it soon.

**Comment by fingolfin** [comment]
Just did a quick experiment and adapted the dockerfile from PR #3 to use `ubuntu:21.10` as base and then followed my suggested edits on that PR. I've also dropped the 32bit support. A quick size comparison:
- Ubuntu 20.04 base image: 65.6MB
- Ubuntu 21.10 base image: 69.9MB
- result of my experiment: 2.68GB (!)
- `gap-docker-master` from 6 days ago: 5.63GB (!!!)

So the *base* image size probably isn't the problem. But the packages we install are. Yet it still saves substantially compared to `gap-docker-master `...

One interesting task would be to figure out how much each individual package we install adds. And then perhaps some could be omitted.

As a first quick test, I checked what happens if I reduce the packages to what is necessary to build just GAP:
```
RUN    apt-get update -qq \
    && apt-get -qq install -y \
            autoconf \
            automake \
            build-essential \
            curl \
            g++ \
            gcc \
            git \
            libgmp-dev \
            libreadline6-dev \
            libtool \
            m4 \
            sudo \
            unzip \
            wget
```
The result already weighs 798MB.

As it turns out, installing the macaulay2 packages adds 1.3 GB (!). Ironically, we install this for the sake of homalg tests, but actually currently the homalg packages don't even specify a `TestFile`, so this is for nothing 😂 

The pari-gp and polymake packages each seem to install about ~450 MB. Singular ~150 MB.

I wonder: is there a way to "extend" the standard GitHub images in a way that leverage layering? What I mean is this: they already use Ubuntu images for their runners, and clearly those are preloaded on the runner machines, so are fast. If we could just layer our images atop those, *and* if GH's infrastructure is able to make use of this (by reusing the already present base image) then we could potentially safe quite a bit, as at least (most of) those 798 MB I just listed are already in there.


One final thought: some of these packages are needed to *build* packages. Some of them are only needed to *run* packages. The latter include some big ones: macaulay2, pari-gp, polymake, singular. Perhaps we should just omit those in the "all-deps" image (resp. in a subimage of that, called "all-build-deps" or so), and then install them only during CI, and only in those CI jobs that really need them...

**Comment by wucas** [comment]
In PR #5 I've tried to make the images as small as I could. But even so given the many combinations we would like this adds up to around 130GB without(!) the doc variant. Where are we going to host that? As far as I can tell ghcr.io offers at most 50GB without additional monthly fees.

**Comment by wucas** [comment]
I stand corrected, I simply didn't read carefully enough and oversaw the complete left side of my monitor where it says "everything free for open source" 🤦🏼‍♂️ my bad.

**Comment by zickgraf** [comment]
I just stumbled across this issue and have some things to add:

* I have opened two PRs which reduce the image size of the existing docker container quite a bit: https://github.com/gap-system/gap-docker-base/pull/32 and https://github.com/gap-system/gap-docker-master/pull/14
* When I added Macaulay2 to the image, the increase in size was only 195MB (https://github.com/gap-system/gap-docker-base/issues/23#issuecomment-544119571). I guess the 1.3 GB @fingolfin has measured include lots of dependencies which also get pulled in by other packages like polymake, or something drastically has changed since then.
* In any case, homalg does not work with current versions of Macaulay2 anyway and we have dropped it from our tests -> feel free to drop it from the container.

**Comment by limakzi** [comment]
@olexandr-konovalov , @fingolfin This repository can be even completely removed, I think.
It had no logic and no source code.

**Comment by fingolfin** [comment]
Yeah, either delete it; or if you archive it please edit the README first to reflect the status (why this is archives / what replaces it)

**Comment by olexandr-konovalov** [comment]
+1 for deletion

**Comment by wilfwilson** [comment]
Yep, delete.
